### DISCUSSION TITLE
DuckDB window function returns tuple instead of value

### DISCUSSION THREAD

**Discussion by angrytongan** [observablehq/framework#1713] [Q&A] [answered]
Not exactly sure where the problem is, so trying here first.

When running a DuckDB window function, the return value of the function is a tuple in Observable Framework:

```sql
WITH
    data AS (
        SELECT 1 AS val
        UNION ALL SELECT 2
        UNION ALL SELECT 3
        UNION ALL SELECT 4
        UNION ALL SELECT 5
        ORDER BY val
    )

SELECT
    val,
    sum(val) OVER (
        ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
    ) AS cumulative
FROM
    data
ORDER BY
    val
```

This query yields:

<img width="474" alt="Screenshot 2024-10-02 at 16 38 53" src="https://github.com/user-attachments/assets/e5f1f737-4c12-4655-a25e-ce4c2582a572">

Casting inside the query does return the right value:

```sql
WITH
    data AS (
        SELECT 1 AS val
        UNION ALL SELECT 2
        UNION ALL SELECT 3
        UNION ALL SELECT 4
        UNION ALL SELECT 5
        ORDER BY val
    )

SELECT
    val,
    CAST(sum(val) OVER (
        ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
    ) AS INTEGER) AS cumulative
FROM
    data
ORDER BY
    val
```

<img width="337" alt="Screenshot 2024-10-02 at 16 40 41" src="https://github.com/user-attachments/assets/972504df-dd48-42b4-934c-bac201fa7022">

Not sure if it's DuckDB or Observable Framework. Would be good if someone could point me in the right direction 👍

**Comment by df-rw** [comment]
Running the (non-cast) query using the duckdb CLI gives:

```
duckdb foo.db
v1.1.1 af39bd0dcf
Enter ".help" for usage hints.
D WITH
      data AS (
          SELECT 1 AS val
          UNION ALL SELECT 2
          UNION ALL SELECT 3
          UNION ALL SELECT 4
          UNION ALL SELECT 5
          ORDER BY val
      )

  SELECT
      val,
      sum(val) OVER (
          ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
      ) AS cumulative
  FROM
      data
  ORDER BY
      val
  ;
┌───────┬────────────┐
│  val  │ cumulative │
│ int32 │   int128   │
├───────┼────────────┤
│     1 │          1 │
│     2 │          3 │
│     3 │          6 │
│     4 │         10 │
│     5 │         15 │
└───────┴────────────┘
```

The output type is a `HUGEINT` ([docs](https://duckdb.org/docs/sql/data_types/numeric.html#integer-types)) which turns into a tuple in Observable Framework. Should types be cast in queries, or is this something Observable Framework should handle?

**Reply by mythmon** [reply]
I'd recommend casting in the query to an appropriately sized integer type, since `sum` will always return HUGEINTs. It wouldn't be correct for DuckDB to automatically cast those to a smaller numeric type.

If you need the extra range of HUGEINT, you could use the title directly and convert it into a JS BigNum, but that would require extra work from what I understand.

**Reply by df-rw** [reply]
Thanks Michael 👍

**Comment by mbostock** [comment, marked as answer]
Yep, this is how Apache Arrow represents large integers. A trivial reproduction of this is:

````md
```sql id=[{n}]
SELECT 11111111100::INT128 AS n
```
````

This defines `n` as a `Arrow.util.BN` instance in the same fashion as if you had written:

```js
const n = new Arrow.util.BN(new Uint16Array([0x19bc, 0x9646, 0x0002, 0, 0, 0, 0, 0]));
```

In order to work with this value as a number, you’ll need to cast it to a number. You can do that in JavaScript as:

```js
Number(n)
```

You can also implicitly cast it to a number _e.g._ as `+n`. 

Alternatively, if you want a bigint, you can go through a string (unfortunately it does not appear that you can coerce directly to BigInt with the current version of Apache Arrow):

```js
BigInt(String(n))
```

Inputs.table doesn’t know about the `BN` class and thus displays its structure as-is rather than casting it to a primitive. As @mythmon said, you can also cast it to a `DOUBLE` in SQL:

```sql
SELECT (11111111100::INT128)::DOUBLE AS n
```

Or, if you know that the value will fit in 32 bits, you could use `INT`:

```sql
SELECT (1234::INT128)::INT AS n
```

There was previously a bug https://github.com/apache/arrow/issues/37920 in how Apache Arrow converted decimals to numbers, but that appears to be fixed now. There’s also https://github.com/duckdb/duckdb-wasm/issues/1703 in DuckDB-Wasm.

Lastly I’ll note that DuckDB-Wasm has a **castBigIntToDouble** query option, which we implicitly default to true in DuckDBClient:

https://github.com/observablehq/framework/blob/e8e590cfb76f290a5461b041097553808c7b04c8/src/client/stdlib/duckdb.js#L168-L170

However, it seems that this doesn’t really function as desired, perhaps because Apache Arrow is returning a `BN` instance instead of a native `BigInt`. So in short, ideally Apache Arrow or DuckDB-Wasm would do this conversion automatically, but they don’t, so currently you’ll have to do it.

**Reply by df-rw** [reply]
Thanks Mike 👍
